src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/LocationRepository.cs

```diff
@@ -16,15 +16,49 @@ public LocationRepository(StatisticsDbContext context) : base(context)
 
         public Dictionary<GeographicLevel, IEnumerable<IObservationalUnit>> GetObservationalUnits(Guid subjectId)
         {
-            var locations = GetLocationsGroupedByGeographicLevel(subjectId);
-            return GetObservationalUnits(locations);
+            var observations = _context.Observation
+                .Where(o => o.SubjectId == subjectId);
+            return GetObservationalUnits(observations);
         }
 
         public Dictionary<GeographicLevel, IEnumerable<IObservationalUnit>> GetObservationalUnits(
             IQueryable<Observation> observations)
         {
-            var locations = GetLocationsGroupedByGeographicLevel(observations);
-            return GetObservationalUnits(locations);
+            var list = observations
+                // NOTE: Not including Location to avoid join for performance
+                .Select(o => new {o.GeographicLevel, o.LocationId})
+                .AsNoTracking()
+                .ToList();
+
+            var geographicLevels = list
+                .Select(elem => elem.GeographicLevel)
+                .Distinct()
+                .ToList();
+
+            // NOTE: Get all locations in a single DB call
+            var locationIds = list
+                .Select(elem => elem.LocationId)
+                .Distinct()
+                .ToArray();
+            var locationsDict = Find(locationIds)
```

Comment:
Just wondering if we are any likely to hit any character limit on SQLServer now when executing this. Following the previous logic, looks like we queried the locations in batches by geographic level, but now without this batching.
Have you tried this with a large subject data file?

Reply:
Did a quick test and I think this will fail at somewhere between 40-50,000 unique location id's. I doubt we'd ever hit a subject file with this many locations.

Reply:
Talked to Ben about this on Slack and agreed to create a ticket to address it: EES-2645.